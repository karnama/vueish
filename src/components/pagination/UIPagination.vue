<template>
    <div :key="page"
         class="flex flex-wrap justify-between items-center max-w-max space-x-2"
         role="navigation"
         aria-label="Pagination navigation">
        <UIButton id="previous-page-button"
                  :disabled="disabled || !hasPrevious"
                  aria-label="Previous Page"
                  class="rotate-90 !p-1"
                  @click="page === 1 ? undefined : setPage(page - 1)"
                  v-html="chevronIcon" />
        <UIButton id="page-1-button"
                  :aria-current="isCurrent = page === 1"
                  :category="isCurrent ? 'primary' : 'default'"
                  :disabled="disabled"
                  :outline="!isCurrent"
                  class="page-button"
                  :aria-label="isCurrent ? 'Current Page, Page 1' : 'Go to page 1'"
                  :title="isCurrent ? 'Current Page, Page 1' : 'Go to page 1'"
                  @click="setPage(1)">
            1
        </UIButton>
        <div v-if="pages[0] !== 2" class="select-none">
            ...
        </div>
        <UIButton v-for="pageNum in pages"
                  :id="`page-${pageNum}-button`"
                  :key="pageNum"
                  :aria-current="isCurrent = pageNum === page"
                  :category="isCurrent ? 'primary' : 'default'"
                  :outline="!isCurrent"
                  class="page-button"
                  :aria-label="isCurrent ? 'Current Page, Page ' + pageNum : 'Go to page ' + pageNum"
                  :title="isCurrent ? 'Current Page, Page ' + pageNum : 'Go to page ' + pageNum"
                  :disabled="disabled"
                  @click="setPage(pageNum)">
            {{ pageNum }}
        </UIButton>
        <div v-if="startPagesFrom + Number(visibleCount) < Number(length)" class="select-none">
            ...
        </div>
        <UIButton v-if="Number(length) > 1"
                  :id="`page-${length}-button`"
                  :aria-current="isCurrent = page === Number(length)"
                  :category="isCurrent ? 'primary' : 'default'"
                  :disabled="disabled"
                  class="page-button"
                  :outline="!isCurrent"
                  :aria-label="isCurrent ? 'Current Page, Page ' + length : 'Go to page ' + length"
                  :title="isCurrent ? 'Current Page, Page ' + length : 'Go to page ' + length"
                  @click="setPage(Number(length))">
            {{ length }}
        </UIButton>
        <UIButton id="next-page-button"
                  :disabled="disabled || !hasNext"
                  aria-label="Next Page"
                  class="rotate-270 !p-1"
                  @click="page === Number(length) ? undefined : setPage(page + 1)"
                  v-html="chevronIcon" />
    </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';
import { useVModel, disabled } from 'composables/input';
import { getIcon } from '@/helpers';
import UIButton from 'components/button/UIButton.vue';

export default defineComponent({
    name: 'UIPagination',

    components: { UIButton },

    props: {
        /**
         * The current page.
         */
        modelValue: {
            type: Number,
            default: 1,
            validator: (val: number) => Number.isInteger(val) && val >= 0
        },

        /**
         * The number of pages.
         */
        length: {
            type: [Number, String],
            required: true,
            validator: (val: number | string) => {
                val = Number(val);
                return !isNaN(val) && Number.isInteger(val) && val > 0;
            }
        },

        /**
         * The total number of page buttons to show.
         */
        visibleCount: {
            type: [Number, String],
            default: 3,
            validator: (val: number | string) => {
                val = Number(val);
                const isValid = !isNaN(val) && Number.isInteger(val) && val >= 0;

                if (isValid && val % 2 !== 1) {
                    throw new Error('visibleCount expected to be an odd number.');
                }

                return isValid;
            }
        },

        disabled
    },

    emits: ['update:modelValue'],

    setup(props) {
        const chevronIcon = getIcon('chevron');

        const page = useVModel<number>(props);
        const hasPrevious = computed(() => page.value > 1);
        const hasNext = computed(() => page.value < Number(props.length));
        const startPagesFrom = computed(() => {
            const sideLength = Math.floor(Number(props.visibleCount) / 2);

            if (page.value <= sideLength + 1) {
                return 2;
            }

            if (page.value >= Number(props.length) - sideLength) {
                return Number(props.length) - Number(props.visibleCount);
            }

            if (Number(props.length) === Number(props.visibleCount)) {
                return page.value - sideLength - 1;
            }
            return page.value - sideLength;
        });
        const pages = computed(() => {
            return Array.from(
                { length: Number(props.visibleCount) },
                (_, index) => index + startPagesFrom.value
            )
                .filter(pageNum => pageNum > 1 && pageNum < Number(props.length));
        });

        const setPage = (pageNum: number) => {
            if (pageNum === page.value) return;

            page.value = pageNum;
        };

        return {
            page,
            hasPrevious,
            startPagesFrom,
            setPage,
            hasNext,
            chevronIcon,
            pages
        };
    }
});
</script>

<style>
.page-button {
    min-width: 40px;
    padding: 0 5px !important;
    height: 38px;
}
</style>
