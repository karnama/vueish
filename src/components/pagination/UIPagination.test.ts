import { mount } from '@vue/test-utils';
import UIPagination from './UIPagination.vue';
import { ref } from 'vue';

const commonProps = {
    modelValue: 5,
    length: 10,
    visibleCount: 3
} as const;

describe('UIPagination', () => {
    it('should render correctly', () => {
        const wrapper = mount(UIPagination, {
            props: commonProps
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display the correct number of buttons', async () => {
        const wrapper = mount(UIPagination, {
            props: commonProps
        });

        expect(wrapper.findAll('button')).toHaveLength(
            // previous and next buttons
            2
            // the buttons
            + commonProps.visibleCount
            // the wrapping end buttons
            + 2
        );

        await wrapper.setProps({ visibleCount: 5 });
        expect(wrapper.findAll('button')).toHaveLength(
            // previous and next buttons
            2
            // the buttons
            + 5
            // the wrapping end buttons
            + 2
        );
    });

    it('should navigate to the previous page on previous button', async () => {
        const wrapper = mount(UIPagination, {
            props: commonProps
        });

        const previousPageBtn = wrapper.find('#previous-page-button');
        await previousPageBtn.trigger('click');

        expect(wrapper.lastEventValue()).toStrictEqual([4]);
        await previousPageBtn.trigger('click');
        await previousPageBtn.trigger('click');
        await previousPageBtn.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([1]);

        await previousPageBtn.trigger('click');
        // it is now disabled so:
        expect(wrapper.lastEventValue()).toStrictEqual([1]);
        expect(wrapper.emitted('update:modelValue')).toHaveLength(4);
    });

    it('should navigate to the next page on next button', async () => {
        const wrapper = mount(UIPagination, {
            props: commonProps
        });

        const nextPageBtn = wrapper.find('#next-page-button');
        await nextPageBtn.trigger('click');

        expect(wrapper.lastEventValue()).toStrictEqual([6]);
        await nextPageBtn.trigger('click');
        await nextPageBtn.trigger('click');
        await nextPageBtn.trigger('click');
        await nextPageBtn.trigger('click');
        expect(wrapper.lastEventValue()).toStrictEqual([10]);

        await nextPageBtn.trigger('click');
        // it is now disabled so:
        expect(wrapper.lastEventValue()).toStrictEqual([10]);
        expect(wrapper.emitted('update:modelValue')).toHaveLength(5);
    });

    it('should handle model-binding correctly', async () => {
        const wrapper = mount({
            components: { UIPagination },
            template: '<UIPagination v-model="value" length="10" visible-count="3" />' +
                '<div id="decrement" @click="value--" />',
            setup: () => {
                return {
                    value: ref(5)
                };
            }
        });

        expect(wrapper.find('#page-5-button').attributes()['title']).toBe('Current Page, Page 5');
        await wrapper.find('#next-page-button').trigger('click');
        expect(wrapper.find('#page-5-button').attributes()['title']).toBe('Go to page 5');
        expect(wrapper.find('#page-6-button').attributes()['title']).toBe('Current Page, Page 6');
        await wrapper.find('#decrement').trigger('click');
        expect(wrapper.find('#page-5-button').attributes()['title']).toBe('Current Page, Page 5');
        expect(wrapper.find('#page-6-button').attributes()['title']).toBe('Go to page 6');
    });

    it('should disable buttons if prop given', () => {
        const wrapper = mount(UIPagination, {
            props: { ...commonProps, disabled: true }
        });

        expect(wrapper.findAll('button[disabled]')).toHaveLength(
            // previous and next buttons
            2
            // the buttons
            + commonProps.visibleCount
            // the wrapping end buttons
            + 2
        );
    });
});
