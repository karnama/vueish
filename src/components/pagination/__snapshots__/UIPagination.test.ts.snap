// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIPagination should render correctly 1`] = `
<div
  aria-label="Pagination navigation"
  class="flex flex-wrap justify-between items-center max-w-max space-x-2"
  role="navigation"
>
  <button
    aria-label="Previous Page"
    class="border-gray-300 text-gray-900 bg-gray-300 disabled:bg-gray-300 ring-gray-600 hover:bg-gray-400 hover:border-gray-400 border px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed rotate-90 !p-1"
    id="previous-page-button"
    type="button"
  >
    <svg
      fill="currentColor"
      height="15px"
      viewBox="0 0 407.437 407.437"
      x="0px"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <polygon
        points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
      />
    </svg>
  </button>
  <button
    aria-current="false"
    aria-label="Go to page 1"
    class="text-gray-600 dark:text-gray-400 shadow-sm border-gray-400 disabled:bg-gray-300disabled:shadow-none ring-gray-600 hover:text-white dark:hover:text-white hover:bg-gray-400 border focus:ring-offset-2 px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed page-button"
    id="page-1-button"
    title="Go to page 1"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label"
      >
        
         1 
        
      </span>
      
    </transition-stub>
  </button>
  <div
    class="select-none"
  >
     ... 
  </div>
  
  <button
    aria-current="false"
    aria-label="Go to page 4"
    class="text-gray-600 dark:text-gray-400 shadow-sm border-gray-400 disabled:bg-gray-300disabled:shadow-none ring-gray-600 hover:text-white dark:hover:text-white hover:bg-gray-400 border focus:ring-offset-2 px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed page-button"
    id="page-4-button"
    title="Go to page 4"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label"
      >
        
        4
        
      </span>
      
    </transition-stub>
  </button>
  <button
    aria-current="true"
    aria-label="Current Page, Page 5"
    class="border-blue-600 text-white bg-blue-600 ring-blue-300 hover:bg-blue-700 hover:border-blue-700 border px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed page-button"
    id="page-5-button"
    title="Current Page, Page 5"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label"
      >
        
        5
        
      </span>
      
    </transition-stub>
  </button>
  <button
    aria-current="false"
    aria-label="Go to page 6"
    class="text-gray-600 dark:text-gray-400 shadow-sm border-gray-400 disabled:bg-gray-300disabled:shadow-none ring-gray-600 hover:text-white dark:hover:text-white hover:bg-gray-400 border focus:ring-offset-2 px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed page-button"
    id="page-6-button"
    title="Go to page 6"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label"
      >
        
        6
        
      </span>
      
    </transition-stub>
  </button>
  
  <div
    class="select-none"
  >
     ... 
  </div>
  <button
    aria-current="false"
    aria-label="Go to page 10"
    class="text-gray-600 dark:text-gray-400 shadow-sm border-gray-400 disabled:bg-gray-300disabled:shadow-none ring-gray-600 hover:text-white dark:hover:text-white hover:bg-gray-400 border focus:ring-offset-2 px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed page-button"
    id="page-10-button"
    title="Go to page 10"
    type="button"
  >
    <transition-stub>
      
      <span
        class="label"
      >
        
        10
        
      </span>
      
    </transition-stub>
  </button>
  <button
    aria-label="Next Page"
    class="border-gray-300 text-gray-900 bg-gray-300 disabled:bg-gray-300 ring-gray-600 hover:bg-gray-400 hover:border-gray-400 border px-3.5 py-2 focus:ring-2 ui-button rounded font-bold text-sm m-0 focus:outline-none ring-0 disabled:cursor-not-allowed transform -rotate-90 !p-1"
    id="next-page-button"
    type="button"
  >
    <svg
      fill="currentColor"
      height="15px"
      viewBox="0 0 407.437 407.437"
      x="0px"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <polygon
        points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
      />
    </svg>
  </button>
</div>
`;
